File: programmers/cpp/programmers_stackqueue_4.cpp
#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
	int answer = 0;
	int sum = 0;
	queue<pair<int, int>> q;

	int i = 0;
	while (1)
	{
		if (i < truck_weights.size())
		{
			if (q.empty() || sum + truck_weights[i] <= weight)
			{
				q.push(make_pair(truck_weights[i], answer));
				sum += truck_weights[i];
				i++;
			}
		}
		if (!q.empty() && q.front().second + bridge_length <= answer + 1)
		{
			sum -= q.front().first;
			q.pop();
		}
		answer++;
		if (i >= truck_weights.size() && q.empty())
			break;
	}
	answer++;
	return answer;
}

File: programmers/python/programmers_stackqueue_4.py
def solution(bridge_length, weight, truck_weights):
	answer = 0
	sum = 0
	i = 0
	bridge = []
	while True:
		if i < len(truck_weights):
			if (len(bridge) == 0 or sum + truck_weights[i] <= weight):
				bridge.append([truck_weights[i], answer])
				sum += truck_weights[i]
				i+= 1
		if len(bridge) != 0 and (bridge[0][1] + bridge_length) <= (answer + 1):
			sum -= bridge[0][0]
			bridge.pop(0)
		answer += 1
		if (i >= len(truck_weights) and len(bridge) == 0):
			break
	answer += 1
	return answer